integration_test/tests/free_and_cancel_tests.rs

```diff
@@ -56,6 +56,82 @@ mod integration {
         }
     }
 
+    unsafe fn exec_direct_default_query(stmt_handle: HStmt) {
+        let mut query: Vec<WideChar> = cstr::to_widechar_vec("SELECT * FROM integration_test.foo");
+        query.push(0);
+        assert_eq!(
+            SqlReturn::SUCCESS,
+            SQLExecDirectW(stmt_handle, query.as_ptr(), SQL_NTS as i32),
+            "{}",
+            get_sql_diagnostics(HandleType::SQL_HANDLE_STMT, stmt_handle as Handle)
+        );
+    }
+
+    /// This test is inspired by the SSIS Query flow. It is altered to be
+    /// more general than that specific flow, with a focus on freeing the
+    /// statement handle after calling SQLBindCol. This flow depends on
+    /// default_setup_connect_and_alloc_stmt.
+    /// After allocating a statement handle, the flow is:
+    ///     - SQLSetStmtAttrW(SQL_ATTR_ROW_BIND_TYPE, SQL_BIND_BY_COLUMN)
+    ///     - SQLSetStmtAttrW(SQL_ATTR_ROW_ARRAY_SIZE, 1000)
+    ///     - SQLExecDirectW(<query>)
+    ///     - <loop: until SQLFetchScroll returns SQL_NO_DATA>
+    ///         - SQLBindCol for each column
+    ///         - SQLFetchScroll
+    ///     - SQLFreeStmt(SQL_UNBIND)
+    ///     - SQLDisconnect
+    ///     - SQLFreeHandle(SQL_HANDLE_DBC)
+    ///     - SQLFreeHandle(SQL_HANDLE_ENV)
+    #[test]
+    fn test_free_stmt_after_bind_col() {
+        let (env_handle, conn_handle, stmt_handle) =
+            default_setup_connect_and_alloc_stmt(AttrOdbcVersion::SQL_OV_ODBC3);
+
+        unsafe {
+            let bind_type = BindType::SQL_BIND_BY_COLUMN as i32;
+            assert_eq!(
+                SqlReturn::SUCCESS,
+                SQLSetStmtAttrW(
+                    stmt_handle,
+                    StatementAttribute::SQL_ATTR_ROW_BIND_TYPE,
+                    bind_type as Pointer,
+                    0,
+                ),
+                "{}",
+                get_sql_diagnostics(HandleType::SQL_HANDLE_STMT, stmt_handle as Handle)
+            );
+
+            let row_array_size = 1000;
+            assert_eq!(
+                SqlReturn::SUCCESS,
+                SQLSetStmtAttrW(
+                    stmt_handle,
+                    StatementAttribute::SQL_ATTR_ROW_ARRAY_SIZE,
+                    row_array_size as Pointer,
+                    0
+                ),
+                "{}",
+                get_sql_diagnostics(HandleType::SQL_HANDLE_STMT, stmt_handle as Handle)
+            );
+
+            exec_direct_default_query(stmt_handle);
+
+            fetch_and_bind_cols(
+                stmt_handle,
+                vec![CDataType::SQL_C_SLONG, CDataType::SQL_C_SBIGINT],
+            );
+
+            assert_eq!(
+                SqlReturn::SUCCESS,
+                SQLFreeStmt(stmt_handle, FreeStmtOption::SQL_UNBIND),
```

Comment:
The actual SSIS trace does `SQLFreeStmt(SQL_CLOSE)`, but I decided not to do that since that is not as interesting as `SQL_UNBIND`. I can change this test or add a second test if either of you thinks it's worth also testing `SQL_CLOSE` after calls to `SQLBindCol`, though those functions shouldn't have anything to do with one another.

Reply:
I agree unbind is more interesting / covers more in integration tests since we already have close 👍

Reply:
True!
You would usually unbind only is you plan to rebind, not if you plan to close your statement/connection.
It might be worth thinking about testing bind/unbind flow specifically if we see a tool doing that.